internal/activities/validate_premis_test.go

```diff
@@ -0,0 +1,154 @@
+	Err error
+}
+
+func (v *fakeValidator) Validate(ctx context.Context, xmlPath, xsdPath string) (string, error) {
+	return v.Msg, v.Err
+}
+
+func (v *fakeValidator) WithMsg(msg string) *fakeValidator {
+	v.Msg = msg
+	return v
+}
+
+func (v *fakeValidator) WithErr(err error) *fakeValidator {
+	v.Err = err
+	return v
+}
+
+func newFakeValidator() *fakeValidator {
+	return &fakeValidator{}
+}
+
+func TestValidatePREMIS(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name      string
+		validator xmlvalidate.XSDValidator
+		params    activities.ValidatePREMISParams
+		want      activities.ValidatePREMISResult
+		wantErr   string
+	}{
+		{
+			name:      "Validates a PREMIS file",
+			validator: xmlvalidate.NewXMLLintValidator(),
+			params: activities.ValidatePREMISParams{
+				Path: fs.NewDir(t, "enduro-test",
+					fs.WithFile("premis.xml", premisXML),
+				).Join("premis.xml"),
+			},
+		},
+		{
+			name:      "Returns a validation failure",
+			validator: newFakeValidator().WithMsg("premis.xml:12: parser error"),
+			params: activities.ValidatePREMISParams{
+				Path: fs.NewDir(t, "enduro-test",
+					fs.WithFile("premis.xml", premisXML),
+				).Join("premis.xml"),
+			},
+			want: activities.ValidatePREMISResult{
+				Failures: []string{"premis.xml does not match expected metadata requirements"},
+			},
+		},
+		{
+			name:      "Returns a file not found failure",
+			validator: newFakeValidator().WithErr(errors.New("file not found")),
+			params: activities.ValidatePREMISParams{
+				Path: fs.NewDir(t, "enduro-test").Join("premis.xml"),
+			},
+			want: activities.ValidatePREMISResult{
+				Failures: []string{"file not found: premis.xml"},
+			},
+		},
+		{
+			name:      "Returns a system error",
+			validator: newFakeValidator().WithErr(errors.New("permission denied: premis.xml")),
+			params: activities.ValidatePREMISParams{
+				Path: fs.NewDir(t, "enduro-test",
+					fs.WithFile("premis.xml", premisXML),
+				).Join("premis.xml"),
+			},
+			wantErr: "permission denied: premis.xml",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			var logbuf bytes.Buffer
+			logger := buflogr.NewWithBuffer(&logbuf)
+
+			ts := &temporalsdk_testsuite.WorkflowTestSuite{}
+			env := ts.NewTestActivityEnvironment()
+			env.SetWorkerOptions(temporalsdk_worker.Options{
+				Interceptors: []temporalsdk_interceptor.WorkerInterceptor{
+					temporal.NewLoggerInterceptor(logger),
+				},
+			})
+			env.RegisterActivityWithOptions(
+				activities.NewValidatePREMIS(tt.validator).Execute,
+				temporalsdk_activity.RegisterOptions{Name: activities.ValidatePREMISName},
+			)
+
+			enc, err := env.ExecuteActivity(activities.ValidatePREMISName, tt.params)
+			if tt.wantErr != "" {
+				if err == nil {
+					t.Errorf("error is nil, expecting: %q", tt.wantErr)
+				} else {
+					assert.ErrorContains(t, err, tt.wantErr)
+				}
+
+				return
+			}
+
+			t.Log(logbuf.String()) // Echo log for debugging.
```

Comment:
Slick!

Reply:
Yeah! It took me a while to figure out how to capture and test the log messages. :sweat_smile: